src/sys/inotify.rs

```diff
@@ -191,52 +185,61 @@ impl Inotify {
     ///
     /// Returns an EINVAL error if the watch descriptor is invalid.
     ///
-    /// For more information see, [inotify_rm_watch(2).](http://man7.org/linux/man-pages/man2/inotify_rm_watch.2.html)
+    /// For more information see, [inotify_rm_watch(2)](http://man7.org/linux/man-pages/man2/inotify_rm_watch.2.html).
     pub fn rm_watch(&self, wd: WatchDescriptor) -> Result<()> {
         let res = unsafe { ffi::inotify_rm_watch(self.fd, wd.wd) };
 
         Errno::result(res).map(drop)
     }
 
-    /// Reads a new event from the inotify file descriptor. This call can either
+    /// Reads ne events from the inotify file descriptor. This call can either
     /// be blocking or non blocking depending on whether IN_NONBLOCK was set at
     /// initialization. 
     /// 
     /// Returns an event if possible. If the call was non blocking and no event 
     /// could be read then the EAGAIN error is returned.
-    pub fn read_event(&self) -> Result<InotifyEvent> {
+    pub fn read_events(&self) -> Result<Vec<InotifyEvent>> {
+        let header_size = size_of::<ffi::inotify_event>();
         let mut buffer = [0u8; 4096];
+        let mut events = Vec::new();
+        let mut offset = 0;
 
-        read(self.fd, &mut buffer)?;
+        let nread = read(self.fd, &mut buffer)?;
 
-        let header_size = std::mem::size_of::<ffi::inotify_event>();
-        assert!(buffer.len() >= header_size);
+        while (nread - offset) >= header_size {
+            let event = unsafe {
+                &*(
+                    buffer
+                        .as_ptr()
+                        .offset(offset as isize) as *const ffi::inotify_event
+                )
+            };
 
-        let event = unsafe {
-            &*(
-                buffer.as_ptr() as *const ffi::inotify_event
-            )
-        };
+            let name = match event.len {
+                0 => None,
+                _ => {
+                    let ptr = unsafe { 
+                        buffer
```

Comment:
This pointer arithmetic is wrong for events past the first because you don't update buffer.

Reply:
Replacing `header_size as isize` by `offset as isize + header_size as isize` should do it right ? Got my arithmetic mixed up. I did not even notice in my tests.